### PR TITLE
PDA Alert Monitor

### LINKED ISSUE #2663: [Feature Request] PDA enhancement purchase - Station Alert Computer for Engi PDAs
Welcome back Deity Link, here's something I think would go over well.  Other stations have portable laptops that function as station alert computers.  Instead of introducing a slew of items to give us a neat toy, give engis, atmos techs, and CEs a PDA purchase that lets them see station alerts from their PDA!

### PR DESCRIPTION
**Pull request by DeityLink**
alert
- Engineer PDA cartridges now grant access to an app that connects to the Station Alert Computers on the Z-level (the same way that the Power Monitor app connects to Power Monitoring Computers).
- The Power Monitor app can now only connect to Power Monitoring Computers on its Z-level.

fixes #2663

### CODE REVIEW

File: code/game/objects/items/devices/PDA/cart.dm [hunk 1]
```diff
@@ -373,6 +378,60 @@ Code:
...
 
 				menu += "</FONT></PRE>"
+		if (53)
+			menu = "<h4><img src=pda_alert.png> Please select an Alert Computer</h4><BR>"
+			alertmonitor = null
+			alertmonitors = list()
+			var/alertcount = 0
+
+
+			for(var/obj/machinery/computer/station_alert/aMon in world)
```
Comment (d3athrow) [inline code/game/objects/items/devices/PDA/cart.dm]: why

File: code/game/objects/items/devices/PDA/cart.dm [hunk 2]
```diff
@@ -373,6 +378,60 @@ Code:
...
+						alertmonitors += aMon
+
+
+			if(!alertcount)
+				menu += "No Alert Computer detected in the vicinity.<BR>"
+			else
+
+				menu += "<FONT SIZE=-1>"
+				var/count = 0
+				for(var/obj/machinery/computer/station_alert/aMon in alertmonitors)
```
Comment (d3athrow) [inline code/game/objects/items/devices/PDA/cart.dm]: are you looping twice

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 3]
```diff
@@ -1044,10 +1042,10 @@ var/global/list/obj/item/device/pda/PDAs = list()
 				if((mode<=9) || (locate(mode) in pda_app_menus))
 					mode = 0
 				else
-					mode = round(mode/10)
-					if(mode==4)//Fix for cartridges. Redirects to hub.
+					mode = round(mode/10)//TODO: fix this shit up
+					if((mode==4) || (mode==5))//Fix for cartridges. Redirects to hub.
 						mode = 0
-					else if(mode >= 40 && mode <= 49)//Fix for cartridges. Redirects to refresh the menu.
+					else if(mode >= 40 && mode <= 53)//Fix for cartridges. Redirects to refresh the menu.
```
Comment (ComicIronic) [inline code/game/objects/items/devices/PDA/PDA.dm]: Can you make this a define?

**Comment by ComicIronic** [comment]
Also is this functionality added to any cartridges?

File: code/game/objects/items/devices/PDA/cart.dm (hunk 1 above)
Comment (DeityLink) [inline code/game/objects/items/devices/PDA/cart.dm]: because I don't know how to isolate the search to a specific Z-level other than by checking the Z-level after finding it.

File: code/game/objects/items/devices/PDA/cart.dm (hunk 2 above)
Comment (DeityLink) [inline code/game/objects/items/devices/PDA/cart.dm]: well yes, but through two different lists, and doing pretty different things each time.

File: code/game/objects/items/devices/PDA/PDA.dm (hunk 3 above)
Comment (DeityLink) [inline code/game/objects/items/devices/PDA/PDA.dm]: I've been pulling my hair over how horrendous cartridges app were designed. All the modes should be defines ideally. But fixing all the shit that's wrong with cartridges is beyond the scope of this PR.

**Comment by DeityLink** [comment]
This functionality is "only" available through cartridges. Any cartridges that give access to the Power Monitor app will also give access to the Alert Monitor app.

File: code/game/objects/items/devices/PDA/cart.dm (hunk 2 above)
Comment (d3athrow) [inline code/game/objects/items/devices/PDA/cart.dm]:
~~~
menu = "No Alert Computer detected in the vicinity"
var/found = 0
-loop-
    if(zlevel)
        if(!found)
              menu = "hreflink"
        else
              menu += "hreflink"
~~~

**Comment by DeityLink** [comment]
Alright, good to go.

File: code/game/objects/items/devices/PDA/cart.dm [hunk 4]
```diff
@@ -374,6 +379,56 @@ Code:
...
+				if(!(aMon.stat & (NOPOWER|BROKEN)))
+					var/turf/T = get_turf(src)
+					if(T.z == aMon.z)//the application may only detect power monitoring computers on its Z-level.
+						if(!found)
+							menu = "<h4><img src=pda_alert.png> Please select an Alert Computer</h4><BR>"
+							found = 1
+							menu += "<FONT SIZE=-1>"
+						alertcount++
+						menu += "<a href='byond://?src=\ref[src];choice=Alert Select;target=[alertcount]'> [aMon] </a><BR>"
+						alertmonitors += aMon
```
Comment (d3athrow) [inline code/game/objects/items/devices/PDA/cart.dm]: To facilitate proper garbage collection i suggest adding the reference in plaintext rather than a hard ref otherwise the GC will have to hard delete it.

Comment (DeityLink) [inline code/game/objects/items/devices/PDA/cart.dm]: In layman's terms? ._.

Comment (d3athrow) [inline code/game/objects/items/devices/PDA/cart.dm]:
~~~
alertmonitors += "\ref[aMon]"

//other code
//interacting with the computer
var/computer/whatever = locate(alertmonitors[thatcomputer])
~~~
